src/engine/effect_functions.asm

```diff
@@ -402,7 +650,356 @@ CreateEnergyCardListFromOpponentDiscardPile: ; 2c2a4 (b:42a4)
 	ret
 ; 0x2c2e0
 
-	INCROM $2c2e0, $2c487
+; returns carry if Deck is empty
+CheckIfDeckIsEmpty: ; 2c2e0 (b:42e0)
+	ld a, DUELVARS_NUMBER_OF_CARDS_NOT_IN_DECK
+	call GetTurnDuelistVariable
+	ldtx hl, NoCardsLeftInTheDeckText
+	cp DECK_SIZE
+	ccf
+	ret
+; 0x2c2ec
+
+; searches through Deck in wDuelTempList looking for
+; a certain card or cards, and prints text depending
+; on whether at least one was found.
+; if none were found, asks the Player whether to look
+; in the Deck anyway, and returns carry if No is selected.
+; uses SEARCHEFFECT_* as input which determines what to search for:
+;	SEARCHEFFECT_CARD_ID = search for card ID in e
+;	SEARCHEFFECT_NIDORAN = search for either NidoranM or NidoranF
+;	SEARCHEFFECT_BASIC_FIGHTING = search for any Basic Fighting Pokemon
+;	SEARCHEFFECT_BASIC_ENERGY = search for any Basic Energy
+;	SEARCHEFFECT_POKEMON = search for any Pokemon card
+; input:
+;	d = SEARCHEFFECT_* constant
+;	e = (optional) card ID to search for in deck
+;	hl = text to print if Deck has card(s)
+; output:
+;	carry set if refused to look at deck
+LookForCardsInDeck: ; 2c2ec (b:42ec)
+	push hl
+	push bc
+	ld a, [wDuelTempList]
+	cp $ff
+	jr z, .none_in_deck
+	ld a, d
+	ld hl, .search_table
+	call JumpToFunctionInTable
+	jr c, .none_in_deck
+	pop bc
+	pop hl
+	call DrawWideTextBox_WaitForInput
+	or a
+	ret
+
+.none_in_deck
+	pop hl
+	call LoadTxRam2
+	pop hl
+	ldtx hl, ThereIsNoInTheDeckText
+	call DrawWideTextBox_WaitForInput
+	ldtx hl, WouldYouLikeToCheckTheDeckText
+	call YesOrNoMenuWithText_SetCursorToYes
+	ret
+; 0x2c317
+
+.search_table
+	dw .SearchDeckForE
+	dw .SearchDeckForNidoran
+	dw .SearchDeckForBasicFighting
+	dw .SearchDeckForBasicEnergy
+	dw .SearchDeckForPokemon
+
+.set_carry ; 2c321 (b:4321)
+	scf
+	ret
+; 0x2c323
+
+; returns carry if no card with
+; same card ID as e is found in Deck
+.SearchDeckForE ; 2c323 (b:4323)
```

Comment:
I would maybe rename this to .SearchDeckForCardID. Maybe a bit less confusing in the first read (also matches the constant name as the other ones)